### ISSUE TITLE
Support `if!` versus `if`

### ISSUE THREAD

**Issue by DavePearce** [Consensys/go-corset#421]
This presents some challenges because `if` performs some kind of inference.  For example, consider this:

```
(defcolumns X Y A)
(defconstraints c1 () (if (eq! X Y) A))
(defconstraints c () (if (eq X Y) A))
```

Both constraints compile down to the same underlying form.  One solution to this issue would be to remove use of `if` altogether from the constraints.  That way, we can reinterpret its meaning as we see fit.

**Comment by DavePearce** [comment]
Another good example:

```
(defcolumns (X :binary@loob) (Y :binary@bool) A)
(defconstraints c1 () (if X A))
(defconstraints c2 () (if Y A))
```

These generate different constraints.  The first is: `if X == 0 then A` whilst the second is `if Y != 0 then A`.

**Comment by DavePearce** [comment]
Fixed as part of #393
